imap_processing/spice/geometry.py

```diff
@@ -149,13 +149,45 @@ def get_spacecraft_spin_phase(
     spin_phase : float or np.ndarray
         Spin phase for the input query times.
     """
+    spin_df = get_spin_data()
+
+    # Combine spin_start_sec and spin_start_subsec to get the spin start
+    # time in seconds. The spin start subseconds are in milliseconds.
+    # TODO: Decide if we should do this calculation in the data itself
+    # or in get_spin_data function.
+    spin_df["spin_start_time"] = (
+        spin_df["spin_start_sec"] + spin_df["spin_start_subsec"] / 1e3
+    )
+
     if isinstance(query_met_times, float):
-        # TODO: call get_spin_data function to get spin data for the
-        # input query times
-        # Here, return a float (dummy implementation)
-        return 0.0  # Replace this with actual logic to calculate spin phase
-
-    # Return an ndarray of the same shape, filled with 0.0 for
-    # now (dummy implementation)
-    # TODO: Replace with actual logic to calculate spin phase
-    return np.array(query_met_times, dtype=float)
+        # calculate spin phase for a single query time
+        mask = spin_df["spin_start_time"] <= query_met_times
+        if mask.any():
+            last_spin = spin_df[mask].iloc[-1]
+            return (query_met_times - last_spin["spin_start_time"]) / last_spin[
+                "spin_period_sec"
+            ]
+        else:
+            raise ValueError(
+                f"No spin data found for this data time: {query_met_times}"
+            )
+
+    # Create an empty array to store spin phase results
+    spin_phases = np.zeros_like(query_met_times)
+
+    # Loop through each query time and find the corresponding spin start time
+    for i, data_time in enumerate(query_met_times):
+        # Find the row where the spin start time is less than or
+        # equal to the query time
+        mask = spin_df["spin_start_time"] <= data_time
+
+        if mask.any():
+            # Get the last spin before or at the query time
+            last_spin = spin_df[mask].iloc[-1]
+            # Calculate spin phase using the formula
+            spin_phases[i] = (data_time - last_spin["spin_start_time"]) / last_spin[
+                "spin_period_sec"
+            ]
+        else:
+            raise ValueError(f"No spin data found for this data time: {data_time}")
```

Comment:
I think this is `searchsorted` to get the indices?
https://numpy.org/doc/stable/reference/generated/numpy.searchsorted.html

```python
# Filter out values above/below, raise, ...?
query_met_times = np.clip(query_met_times, spin_df["spin_start_time"][0], spin_df["spin_start_time"][-1])
last_spin_indices = np.searchsorted(spin_df["spin_start_time"], query_met_times)
spin_df_indexed = spin_df.iloc[last_spin_indices]
spin_phases = (query_met_times - spin_df_indexed["spin_start_time"]) / spin_df_indexed["spin_period_sec"]
```

Reply:
Do you need to multiply by `2*pi` here? Are we expecting spin phase in radians?

Reply:
The suggested use of `np.clip` scares me because it will silently modify the query times and cause this function to return incorrect spin phase rather than raising an exception.

Reply:
I was just reading on those two and I don't understand `np.clip` but `np.searchsorted` seems like a good approach. I am still reading though :)

Reply:
> The suggested use of np.clip scares me because it will silently modify the query times and cause this function to return incorrect spin phase rather than raising an exception.

Fair enough, we should never encounter that situation anyways, so I am fine with raising an error instead of removing the input requests (probably smarter for shapes of arrays as well!). The issue is you do still need to add some check for out of bounds because I think `searchsorted` will return 0 even if it is outside of the leftmost value.

Reply:
Look like np.clips does help in some things. Eg.
```
>>> x = np.searchsorted(df['a'], [0, 17, 19])
>>> x
array([0, 5, 5])
>>> np.clip(x,0, len(df['a'])-1)
array([0, 4, 4])
```